### PR TITLE
Week3

### PR DESCRIPTION
**Pull request by yuzhiliu**
This version added an extra layer on top of the trained model to preprocess the question asked so that the input question is always valid. This makes the code much more flexible than the original one.

### CODE REVIEW

File: graphqa/predict.py [hunk 1]
```diff
@@ -92,13 +93,46 @@ def download_model(args):
...
+    return dot_product / (norm_a * norm_b)
+
+
+def extract_one_bert(query, questions):
+    """
+    User BERT to find the most similar question from the questions. This
+    can also be realized by fuzzywuzzy.
+    BERT Server needs to be started first. See start-bert-server.sh.
+    """
+    bc = BertClient(port=5555, port_out=5556)
```
Comment (JMHOO) [inline graphqa/predict.py:122]: Have you imported `BertClient`?

File: graphqa/util.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def pretokenize_english(text):
     text = pretokenize_general(text)
     #for p in ENGLISH_PUNCTUATION:
     #    text = text.replace(p, f" {p} ")
-    table = string.maketrans({p: f" {p} " for p in ENGLISH_PUNCTUATION})
+    table = str.maketrans({p: f" {p} " for p in ENGLISH_PUNCTUATION})
```
Comment (JMHOO) [inline graphqa/util.py]: `str.maketrans` can be done in Global level

File: graphqa/predict.py (hunk 1 above)
Comment (JMHOO) [reply graphqa/predict.py:122]: Perhaps, should initialize `BertClient` as global, then call it when needed.

Comment (yuzhiliu) [reply graphqa/predict.py:122]: The BertClient import is now fixed. Indeed, the BertClient call does not belong here. This is a simple dirty way of testing BERT encoder. If one wants to use BERT for encoding the question and use BertServer to serve the model, one may also need to specify the ip addres of the BertServer and tweak some other configurations. I will leave these for later. Also, the function "extract_one_bert" is commented out and not being called in the current code.

File: graphqa/util.py (hunk 2 above)
Comment (yuzhiliu) [reply graphqa/util.py]: Now put str.maketrans at the top of the script. Also created test scripts to test the related fucntions.

**Comment by JMHOO** [comment]
Looks good

**Review by JMHOO — APPROVED** [review]

**Review by JMHOO — COMMENTED** [review]
Looks good

**Pull request merged**